### ISSUE TITLE
Vulnerability to stack smashing attacks

### ISSUE THREAD

**Issue by rtfeldman** [roc-lang/roc#806]
## Summary

On most systems there's a [security vulnerability](https://lwn.net/Articles/725832/) for stack allocations over 4096B. It's a buffer overrun attack which can be used to read and/or overwrite arbitrary heap data. We want to prevent this in order to give Roc strong security guarantees (e.g. installing an editor plugin that is 100% Roc must not even potentially result in arbitrary heap data being changed).

Fortunately, it seems we can prevent this by making sure whenever we instantiate a stack-allocated value (e.g. a record) in generated Roc code, we always write into its memory addresses from highest address to lowest. (This includes tag payloads, tag discriminants, and host-allocated closure data, since hosts are encouraged to allocate that data using `alloca`!)

We could also consider using heap allocations for all values over 4096B. This doesn't seem strictly necessary for security as long as we can be sure we're always writing into stack addresses from highest to lowest (although it would have the separate benefit of making stack overflows less likely in practice). That said, if LLVM might reorder our writes, it might indeed be necessary to prevent stack smashing attacks. (If we did this, we would no longer needed to worry about order of writes.)

## Background

Typically the heap begins at a memory address that's a fairly low number (let's say `2048` for example) and then "grows" upwards - which is to say, new heap allocations have memory addresses that are higher numbers than previous heap allocations' memory addresses.

In contrast, the stack typically begins at a memory address that's a high number (let's say `98765432104` for example) and then "grows" downward - which is to say, new stack allocations have memory addresses that are lower numbers than previous stack allocations' memory addresses.

Usually this works out fine, but there are some ways things can go wrong.

For example, if the heap keeps expanding, eventually it will try to use addresses that overlap with the stack. My understanding is that this will cause `mmap` (which `malloc` uses to bring in new memory into the process's virtual address space) to fail, which in turn will cause `malloc` to fail, because the OS will say "I don't have any remaining un-reserved memory pages I can reserve for you" - because it knows the remaining pages are all reserved for the stack. It's undesirable when this happens, but there's no security vulnerability there.

Stack allocations (e.g. with `alloca`) don't call `malloc` or `mmap` - they just move the stack pointer and start happily using whatever memory happens to be there, under the assumption that whatever it's pointing to now is totally safe to use. One of the reasons stack allocations are super fast is that there are no checks here! Of course, that means there's no check for "running out of stack space" by default - if the stack pointer is 16B below what's supposed to be the end of the stack, and you `alloca` 32 bytes, then you'll theoretically have a pointer to something that is partially located inside the heap. Whoops!

A way to prevent this from happening is to put a [guard page](https://blogs.gnome.org/markmc/2005/05/11/stack-guard-page/) right where you want the stack to end. The guard page is a 4096B block of memory addresses which are configured in the OS to have the rule that if any code ever tries to write to (or read from) that memory range, it should immediately segfault. This way, if we `alloca` a 32B value when there are only 16B left in the stack, we'll overflow into the guard page, and as soon as we try to read from or write to that area, instead of overlapping with heap memory, we'll get a segfault. Segfaults are undesirable, but much better than overwriting arbitrary heap data!

According to [the article linked above](https://lwn.net/Articles/725832/), it's apparently standard these days for Linux distributions to have between 4096B and 1MB worth of guard pages - and presumably other OSes have similar defaults.

## The Attack

A determined attacker can get around the guard page by making a very large stack allocation. Say I define a 9000B record, and then I'm careful to only read from/write to the very last few fields in it. The guard page only segfaults if I actually read from or write to one or more of its bytes, so as long as I make sure never to touch the fields that are allocated in the first 4096B of my record, I'll never trigger the segfault. Meanwhile, I can now use the later bytes in that record to read heap data and even overwrite it! At this point I have a full-on buffer overflow attack.

In C (and presumably also Rust), this attack can be performed by allocating a large struct on the stack, with undefined values in its first 4096B, and then only ever touching its later fields. If `alloca` had a cap on how big individual allocations could be, in theory (as far as I know) an attacker could get around that restriction by allocating a bunch of smaller variables (e.g `int a1, b1, c1, d1, e1, f1, g1, ..., y99, z99;`) and then only ever accessing the later ones. (To prevent the first ones from being eliminated as dead code, they could write code which accesses them only after the nefarious heap tampering was complete, or they could put the accesses behind a conditional they know will always fail.)

Adding multiple consecutive guard pages doesn't rule out the attack, because there are still a finite number of them; if there are 1MB of guard pages, the attacker can allocate 1MB worth of untouched stack space and then however many extra bytes are desired for the attack.

## Preventing the attack in Roc

There's no way to declare undefined values in Roc code. Whenever our code gen allocates anything, by design the very next thing it does is to generate a write into that memory - to make sure it's not undefined. That write would cause the guard page to segfault for any stack allocation under 4096B.

The only way I could think of that an allocation over 4096B could be used to successfully execute a buffer overrun in Roc is if we happened to populate stack-allocated values containing multiple values (e.g. record fields, tag payloads, tag discriminants, and captured data in closures - possibly others too!) starting from the lower memory addresses rather than the higher ones.

For example, let's say someone requests a 12000B record. As usual, the next thing Roc code gen would do would be to write into the memory of each of those fields to populate them, leaving none undefined. If we happened to write into the lower memory addresses first, then we'd still write nearly 8000B of data into the heap before hitting the guard page and segfaulting. Now pretend that instead of 12000B it's a much bigger number, and you can imagine how an attacker could use this to overwrite theoretically as much as the entire heap.

Writing into heap memory immediately before causing a segfault might sound useless ("the program segfaulted, so who cares?") but hosts are supposed to be designed to recover from stack overflows (by trapping these segfaults) and to recover from them so the program can continue even after the stack overflowed (e.g. if an editor plugin overflows the stack, we don't want to crash the whole editor) - so an attacker being able to overwrite host memory in this way could theoretically be used in a real attack.

Fortunately, this seems to be solvable by making sure we always write into record and tag fields from highest memory address to lowest. (We should verify that LLVM doesn't potentially reorder these in undesirable ways.) That way, if our allocation is so big that we are definitely going to encounter a guard page, we'll encounter it and segfault before there is any possibility of data being written past the guard page.

Separately, we could consider using heap allocations for all values over 4096B in size (e.g. "huge records go on the heap"), but this doesn't seem strictly necessary for security unless LLVM might reorder our writes. That said, taking this measure would mean that we didn't have to worry about order of writes. (Security implications aside, it could separately make stack overflows less likely in practice, which has its own benefits.)
